lang/c++/impl/Compiler.cc

```diff
@@ -327,6 +338,50 @@ static NodePtr makeRecordNode(const Entity& e,
         fieldValues, fieldNames, defaultValues));
 }
 
+static LogicalType makeLogicalType(const Entity& e, const Object& m) {
+    if (!containsField(e, m, "logicalType")) {
+        return LogicalType(LogicalType::NONE);
+    }
+
+    const std::string& typeField = getStringField(e, m, "logicalType");
+
+    if (typeField == "decimal") {
+        LogicalType decimalType(LogicalType::DECIMAL);
+        try {
+            decimalType.setPrecision(getLongField(e, m, "precision"));
+            if (containsField(e, m, "scale")) {
+                decimalType.setScale(getLongField(e, m, "scale"));
+            }
+        } catch (Exception& ex) {
+            // If any part of the logical type is malformed, per the standard we
+            // must ignore the whole attribute.
+            return LogicalType(LogicalType::NONE);
+        }
+        return decimalType;
+    }
+    if (typeField == "date") {
```

Comment:
I think this will save some code and become more readable if we do something like:

    LogicalType::Type t = LogicalType::NONE;
    if (typeField == "date") {
        t = LogicalType::DATE;
    else if (typeField == "time-millis") {
        t = LogicalType::TIME_MILLIS;
   else if ... {
   }
   return LogicalType(t);